NPE when fetching source from an index with _source disabled

**Elasticsearch version**: 2.4.0

**Plugins installed**: [Marvel,Head,License,Delete-by-query]

**JVM version**: 1.8.0_91

**OS version**: linux ubuntu 14.04 - 4.5.7

**Description of the problem including expected versus actual behavior**:

I have 8 indexes, called url-0,url-1,...,url-8 with around 1B documents.
The template for each index is 

    {
        "template" : "url-*",
        "settings" : {            
            "number_of_shards"   : "1",
            "number_of_replicas" : "0",
            "refresh_interval"   : "180s"
        },
        "mappings" : {
            "url" : {
                "_all" : {"enabled" : false},
                "_timestamp" : {"enabled" : false},
                "_source": {"enabled": false},
                "_ttl" : { "enabled" : false },
                "properties" : {
                    "url" : {"type" : "string", "index": "not_analyzed", "store" : true},
                    "domain" : {"type": "string", "index" : "not_analyzed", "store" : false},
                    "scheme" : {"type": "string", "index" : "not_analyzed", "store" : false},
                    "parsedUrl" : {"type": "string", "index": "analyzed", "store": false},
                    "workerIndex": {"type": "integer", "store" : false},
                    "error": {"type": "boolean", "store" : false},
                    "banned": {"type": "boolean", "store" : false},
                    "timestamp": {"type": "date", "store" : true},
                    "depth": {"type": "integer", "store" : true},
                    "processed": {"type": "boolean", "store" : false}

                }
            }
        }

When I execute the query

`curl -XPOST localhost:9200/url/url/_search?search_type=scan&scroll=5m&size=50&_source=url&preference=_shards:0;_prefer_node:XqfWWAGFRdCPo98qIeFfGg`

I get

    {

        "_scroll_id": "c2Nhbjs4OzMyNzUyNzpYcWZXV0FHRlJkQ1BvOThxSWVGZkdnOzQ3MTkzODplTDZZa3AxSlRVQ3lHd1FMTjR3T3NnOzgxMzY4ODpwSHVGaVJRM1JSYTIwaWRxNTlFbXNBOzgxMzY4NzpwSHVGaVJRM1JSYTIwaWRxNTlFbXNBOzgxMzY4NjpwSHVGaVJRM1JSYTIwaWRxNTlFbXNBOzQ3MTkzOTplTDZZa3AxSlRVQ3lHd1FMTjR3T3NnOzMxNDc1NDpqbUdTWGRmWVRTdWFIT1FFSmZ1WVN3OzgxMzY4OTpwSHVGaVJRM1JSYTIwaWRxNTlFbXNBOzE7dG90YWxfaGl0czoxMDEzOTAyNTk1Ow==",
        "took": 8221,
        "timed_out": false,
        "_shards": {
            "total": 8,
            "successful": 8,
            "failed": 0
        },
        "hits": {
            "total": 1013902595,
            "max_score": 0,
            "hits": [ ]
        }
    }

but

If I query for the scrollId I get an error and a NPE on the server.

    culr -XPOST localhost:9200/_search/scroll?scroll=1m -d '
    {
      "scroll_id": "c2Nhbjs4OzMyNzUyNzpYcWZXV0FHRlJkQ1BvOThxSWVGZkdnOzQ3MTkzODplTDZZa3AxSlRVQ3lHd1FMTjR3T3NnOzgxMzY4ODpwSHVGaVJRM1JSYTIwaWRxNTlFbXNBOzgxMzY4NzpwSHVGaVJRM1JSYTIwaWRxNTlFbXNBOzgxMzY4NjpwSHVGaVJRM1JSYTIwaWRxNTlFbXNBOzQ3MTkzOTplTDZZa3AxSlRVQ3lHd1FMTjR3T3NnOzMxNDc1NDpqbUdTWGRmWVRTdWFIT1FFSmZ1WVN3OzgxMzY4OTpwSHVGaVJRM1JSYTIwaWRxNTlFbXNBOzE7dG90YWxfaGl0czoxMDEzOTAyNTk1Ow"
    }'

result:

    {

        "_scroll_id": "c2NhbjswOzE7dG90YWxfaGl0czoxMDEzOTAyNTk1Ow==",
        "took": 105,
        "timed_out": false,
        "_shards": {
            "total": 8,
            "successful": 0,
            "failed": 8,
            "failures": [
                {
                    "shard": -1,
                    "index": null,
                    "reason": {
                        "type": "null_pointer_exception",
                        "reason": null
                    }
                }
            ]
        },
        "hits": {
            "total": 1013902595,
            "max_score": 0,
            "hits": [ ]
        }

    }


And in the logs:

    RemoteTransportException[[es-1][][indices:data/read/search[phase/scan/scroll]]]; nested: NullPointerException;
    Caused by: java.lang.NullPointerException
            at org.elasticsearch.search.fetch.source.FetchSourceSubPhase.hitExecute(FetchSourceSubPhase.java:79)
            at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:188)
            at org.elasticsearch.search.SearchService.executeScan(SearchService.java:342)

Maybe it is a problem related to the **_source** parameter in the query and the fact that the _source field is disabled in the mapping. If I remove the **_source** field from the query, it works fine.

PS: The above query was originated by Spark with the elasticsearch-spark_2.11 (v.2.4.0) library.

--- a/core/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourceSubPhase.java
+++ b/core/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourceSubPhase.java
@@ -36,14 +36,16 @@ public final class FetchSourceSubPhase implements FetchSubPhase {
             return;
         }
         SourceLookup source = context.lookup().source();
-        if (source.internalSourceRef() == null) {
-            return; // source disabled in the mapping
-        }
         FetchSourceContext fetchSourceContext = context.fetchSourceContext();
         assert fetchSourceContext.fetchSource();
         if (fetchSourceContext.includes().length == 0 && fetchSourceContext.excludes().length == 0) {
             hitContext.hit().sourceRef(source.internalSourceRef());
             return;
+        }
+
+        if (source.internalSourceRef() == null) {
+            throw new IllegalArgumentException("unable to fetch fields from _source field: _source is disabled in the mappings " +
+                    "for index [" + context.indexShard().shardId().getIndexName() + "]");
         }
 
         Object value = source.filter(fetchSourceContext.includes(), fetchSourceContext.excludes());

--- a/core/src/test/java/org/elasticsearch/search/fetch/subphase/FetchSourceSubPhaseTests.java
+++ b/core/src/test/java/org/elasticsearch/search/fetch/subphase/FetchSourceSubPhaseTests.java
@@ -23,6 +23,8 @@ import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.bytes.BytesReference;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.search.fetch.FetchSubPhase;
 import org.elasticsearch.search.internal.InternalSearchHit;
 import org.elasticsearch.search.internal.SearchContext;
@@ -33,36 +35,10 @@ import org.elasticsearch.test.TestSearchContext;
 import java.io.IOException;
 import java.util.Collections;
 
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
 public class FetchSourceSubPhaseTests extends ESTestCase {
-
-    static class FetchSourceSubPhaseTestSearchContext extends TestSearchContext {
-
-        FetchSourceContext context;
-        BytesReference source;
-
-        FetchSourceSubPhaseTestSearchContext(FetchSourceContext context, BytesReference source) {
-            super(null);
-            this.context = context;
-            this.source = source;
-        }
-
-        @Override
-        public boolean sourceRequested() {
-            return context != null && context.fetchSource();
-        }
-
-        @Override
-        public FetchSourceContext fetchSourceContext() {
-            return context;
-        }
-
-        @Override
-        public SearchLookup lookup() {
-            SearchLookup lookup = super.lookup();
-            lookup.source().setSource(source);
-            return lookup;
-        }
-    }
 
     public void testFetchSource() throws IOException {
         XContentBuilder source = XContentFactory.jsonBuilder().startObject()
@@ -109,11 +85,14 @@ public class FetchSourceSubPhaseTests extends ESTestCase {
         hitContext = hitExecute(null, false, null, null);
         assertNull(hitContext.hit().sourceAsMap());
 
-        hitContext = hitExecute(null, true, "field1", null);
-        assertNull(hitContext.hit().sourceAsMap());
+        IllegalArgumentException exception = expectThrows(IllegalArgumentException.class, () -> hitExecute(null, true, "field1", null));
+        assertEquals("unable to fetch fields from _source field: _source is disabled in the mappings " +
+                "for index [index]", exception.getMessage());
 
-        hitContext = hitExecuteMultiple(null, true, new String[]{"*"}, new String[]{"field2"});
-        assertNull(hitContext.hit().sourceAsMap());
+        exception = expectThrows(IllegalArgumentException.class,
+                () -> hitExecuteMultiple(null, true, new String[]{"*"}, new String[]{"field2"}));
+        assertEquals("unable to fetch fields from _source field: _source is disabled in the mappings " +
+                "for index [index]", exception.getMessage());
     }
 
     private FetchSubPhase.HitContext hitExecute(XContentBuilder source, boolean fetchSource, String include, String exclude) {
@@ -130,5 +109,41 @@ public class FetchSourceSubPhaseTests extends ESTestCase {
         FetchSourceSubPhase phase = new FetchSourceSubPhase();
         phase.hitExecute(searchContext, hitContext);
         return hitContext;
+    }
+
+    private static class FetchSourceSubPhaseTestSearchContext extends TestSearchContext {
+        final FetchSourceContext context;
+        final BytesReference source;
+        final IndexShard indexShard;
+
+        FetchSourceSubPhaseTestSearchContext(FetchSourceContext context, BytesReference source) {
+            super(null);
+            this.context = context;
+            this.source = source;
+            this.indexShard = mock(IndexShard.class);
+            when(indexShard.shardId()).thenReturn(new ShardId("index", "index", 1));
+        }
+
+        @Override
+        public boolean sourceRequested() {
+            return context != null && context.fetchSource();
+        }
+
+        @Override
+        public FetchSourceContext fetchSourceContext() {
+            return context;
+        }
+
+        @Override
+        public SearchLookup lookup() {
+            SearchLookup lookup = super.lookup();
+            lookup.source().setSource(source);
+            return lookup;
+        }
+
+        @Override
+        public IndexShard indexShard() {
+            return indexShard;
+        }
     }
 }

--- a/core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java
+++ b/core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java
@@ -19,7 +19,6 @@
 package org.elasticsearch.search.fetch.subphase.highlight;
 
 import com.carrotsearch.randomizedtesting.generators.RandomPicks;
-
 import org.apache.lucene.search.join.ScoreMode;
 import org.elasticsearch.Version;
 import org.elasticsearch.action.index.IndexRequestBuilder;
@@ -50,8 +49,8 @@ import org.hamcrest.Matcher;
 import org.hamcrest.Matchers;
 
 import java.io.IOException;
-import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
 
@@ -96,7 +95,7 @@ public class HighlighterSearchIT extends ESIntegTestCase {
 
     @Override
     protected Collection<Class<? extends Plugin>> nodePlugins() {
-        return Arrays.asList(InternalSettingsPlugin.class);
+        return Collections.singletonList(InternalSettingsPlugin.class);
     }
 
     public void testHighlightingWithWildcardName() throws IOException {

--- a/core/src/test/java/org/elasticsearch/search/fields/SearchFieldsIT.java
+++ b/core/src/test/java/org/elasticsearch/search/fields/SearchFieldsIT.java
@@ -80,7 +80,7 @@ public class SearchFieldsIT extends ESIntegTestCase {
 
     @Override
     protected Collection<Class<? extends Plugin>> nodePlugins() {
-        return Arrays.asList(CustomScriptPlugin.class);
+        return Collections.singletonList(CustomScriptPlugin.class);
     }
 
     public static class CustomScriptPlugin extends MockScriptPlugin {

